native/dsteqr.go

```diff
@@ -0,0 +1,386 @@
+			mm := m - l + 1
+			impl.Dlasr(blas.Right, lapack.Variable, lapack.Backward,
+				n, mm, work[l:], work[n-1+l:], z[l:], ldz)
+		}
+		d[l] -= p
+		e[l] = g
+		goto Fourty
+
+		// Eigenvalue found.
+	Eighty:
+		d[l] = p
+		l++
+		if l <= lend {
+			goto Fourty
+		}
+		goto OneFourty
+	} else {
+		// QR Iteration.
+		// Look for small superdiagonal element.
+	Ninety:
+		if l != lend {
+			lendp1 := lend + 1
+			for m = l; m >= lendp1; m-- {
+				v := math.Abs(e[m-1])
+				test := v * v
+				if test <= (eps2*math.Abs(d[m])*math.Abs(d[m-1]) + safmin) {
+					goto OneTen
+				}
+			}
+		}
+		m = lend
+	OneTen:
+		if m > lend {
+			e[m-1] = 0
+		}
+		p = d[l]
+		if m == l {
+			goto OneThirty
+		}
+		// If remaining matrix is 2×2, use Dlae2 to compute its eigenvalues.
+		if m == l-1 {
+			if icompz > 0 {
+				rt1, rt2, c, s = impl.Dlaev2(d[l-1], e[l-1], d[l])
+				work[m] = c
+				work[n-1+m] = s
+				impl.Dlasr(blas.Right, lapack.Variable, lapack.Forward,
+					n, 2, work[m:], work[n-1+m:], z[l-1:], ldz)
+			} else {
+				rt1, rt2 = impl.Dlae2(d[l-1], e[l-1], d[l])
+			}
+			d[l-1] = rt1
+			d[l] = rt2
+			e[l-1] = 0
+			l -= 2
+			if l >= lend {
+				goto Ninety
+			}
+			goto OneFourty
+		}
+		if jtot == nmaxit {
+			goto OneFourty
+		}
+		jtot++
+
+		// Form shift.
+		g = (d[l-1] - p) / (2 * e[l-1])
+		r = impl.Dlapy2(g, 1)
+		g = d[m] - p + (e[l-1])/(g+math.Copysign(r, g))
+
+		s = 1
+		c = 1
+		p = 0
+
+		// Inner loop.
+		lm1 = l - 1
+		for i := m; i <= lm1; i++ {
+			f := s * e[i]
+			b := c * e[i]
+			c, s, r = impl.Dlartg(g, f)
+			if i != m {
+				e[i-1] = r
+			}
+			g = d[i] - p
+			r = (d[i+1]-g)*s + 2*c*b
+			p = s * r
+			d[i] = g + p
+			g = c*r - b
+
+			// If eigenvectors are desired, then save rotations.
+			if icompz > 0 {
+				work[i] = c
+				work[n-1+i] = s
+			}
+		}
+		// If eigenvectors are desired, then apply saved rotations.
+		if icompz > 0 {
+			mm := l - m + 1
+			impl.Dlasr(blas.Right, lapack.Variable, lapack.Forward,
+				n, mm, work[m:], work[n-1+m:], z[m:], ldz)
+		}
+		d[l] -= p
+		e[lm1] = g
+		goto Ninety
+
+		// Eigenvalue found
+	OneThirty:
+		d[l] = p
+		l--
+		if l >= lend {
+			goto Ninety
+		}
+		goto OneFourty
+	}
+OneFourty:
+	// Undo scaling if necessary.
+	if iscale == 1 {
+		impl.Dlascl(lapack.General, 0, 0, ssfmax, anorm, lendsv-lsv+1, 1, d[lsv:], n)
+		impl.Dlascl(lapack.General, 0, 0, ssfmax, anorm, lendsv-lsv, 1, e[lsv:], n)
+	} else if iscale == 2 {
+		impl.Dlascl(lapack.General, 0, 0, ssfmin, anorm, lendsv-lsv, 1, e[lsv:], n)
+		impl.Dlascl(lapack.General, 0, 0, ssfmin, anorm, lendsv-lsv, 1, e[lsv:], n)
+	}
+
+	// Check for no convergence to an eigenvalue after a total of n*maxit iterations.
+	if jtot < nmaxit {
+		goto Ten
+	}
+	for i := 0; i < n-1; i++ {
+		if e[i] != 0 {
+			return false
+		}
+	}
+OneSixty:
+	// Order eigenvalues and eigenvectors.
+	if icompz == 0 {
+		impl.Dlasrt(lapack.SortIncreasing, n, d)
+	} else {
+		// TODO(btracey): replace this sort with a call to sort.Sort.
```

Comment:
Mostly because insertion sort isn't likely to be the fastest, but comment changed.